examples/read_write_mutex/source/app.d

```diff
@@ -0,0 +1,89 @@
+
+import vibe.d;
+
+interface MyBlockingRestInterface {
+	@path("/")
+	int getIndex();
+}
+
+class RestInterfaceImplementation : MyBlockingRestInterface {
+	private {
+		TaskReadWriteMutex m_mutex;
+	}
+
+	this()
+	{
+		m_mutex = new TaskReadWriteMutex(TaskReadWriteMutex.Policy.PREFER_WRITERS);
+	}
+
+	//This method naively simulates access to a shared resource protected
+	//by a TaskReadWriteMutex. Read operations are performed frequently (95%) 
+	//and finish quickly (1ms wait time), while write operations occur rarely
+	//(5%) and take much longer to process (50ms wait time).
+	int getIndex()
+	{
+	    import std.random;
+	    import std.stdio;
+	    auto i = uniform(0,20);
+	    if (i == 1) //In a rare case, lock for writing
+	    {
+	        synchronized(m_mutex.writer)
+	        {
+	        	version(PrintRequests)
+	        		writeln("Blocked for writing!");
+	        	//Simulate a slow operation
+	            vibe.core.core.sleep(50.msecs);
+	        }
+	    }
+	    else //More regularly, lock for reading
+	    {
+	        synchronized(m_mutex.reader)
+	        {
+	        	version(PrintRequests)
+	        		writeln("Blocked for reading!");
+	        	//Simulate a faster operation
+	            vibe.core.core.sleep(1.msecs);
+	        }
+	    }
+	    return 0;
+	}
+}
+
+shared static this()
+{
+	auto routes = new URLRouter;
+	registerRestInterface(routes, new RestInterfaceImplementation());
+
+	auto settings = new HTTPServerSettings();
+	settings.port = 8080;
+	settings.options = HTTPServerOption.parseURL | HTTPServerOption.distribute;
+	settings.bindAddresses = ["::1", "127.0.0.1"];
+
+	listenHTTP(settings, routes);
+
+	//Wait for a couple of seconds for the server to be initialized properly and then start
+	//multiple concurrent threads that simultaneously start queries on the rest interface defined above.
+	setTimer(5.seconds, {
+        import core.atomic;
+
+		scope(exit)
+			exitEventLoop(true);
+
+        auto runningTasks = new shared(int)(0);
+        runWorkerTaskDist(function(typeof(runningTasks) runningTasks) {
+                atomicOp!"+="(*runningTasks,1);
+                scope(exit)
+                    atomicOp!"-="(*runningTasks,1);
+
+                auto api = new RestInterfaceClient!MyBlockingRestInterface("http://127.0.0.1:8080");
+                for (int i = 0; i < 1000; ++i)
+                    api.getIndex();
+
+            }, runningTasks);
+         
+        //Join all worker tasks. Currently, it's not possible to join tasks from non-vibe threads.
+        do 
+            sleep(3.seconds);
+        while(atomicLoad(*runningTasks) > 0);        
```

Comment:
Forgive me if I'm overlooking something; I'll try to explain which specific case I'm afraid of:

```{D}
synchronized (s_taskMutex)
        while (s_taskCounter > 0)
            s_taskCondition.wait();
```
This code basically comes down to check -> wait -> check -> wait etc.
The way `TaskCondition.wait()` is currently implemented, there seems (to me at least) to be a possibility, that the state of the condition changes *between* **check** and **wait** in a concurrent setup. This is the problem I also see if `Future` would be implemented with a `TaskCondition`.
If the check leads to a waiting operation, but the notify() has already propagated before the method is executed, `TaskCondition.wait()` will internally fetch the **new** emitCount and produce a deadlock.

Whereas with `ManualEvent`, one could write:
```{D}
auto count = s_event.emitCount;
while (atomicLoad(s_taskCounter) > 0) //Atomic since we don't use locking here
    count = s_event.wait(count);
```
`ManualEvent.emitCount()` seems to perform an `atomicLoad()` operation and the resulting memory fences should guarantee that the count is always extracted before the check is performed. This should make deadlocks impossible.

Again, forgive me if I'm nitpicking or misunderstanding something, but I have personally already been burned by example code that produced deadlocks with some very small propability. Because of that, I think it is important to promote *correct* code even if it might not be absolutely necessary for a simple example, as those kinds of bugs are a pain to hunt down in more complex systems.


> I'm still reluctant to commit to a specific shared approach until it's clear that this will stay valid in the long term

I completely understand that, the whole matter seems rather perplexing to me.